### ISSUE TITLE
Idea : C++ Mobs.

### ISSUE THREAD

**Issue by C1ffisme** [minetest/minetest#2291]
You've seen the mods. Just download one and see the bugs or lag. If we could have the engine running the behavior of mobs, I think that they would go faster. We could make a function like minetest.register_mob(), then all you would have to do for an animals mod is use the (hopefully many) features in the function, and make models and textures (and maybe some sounds). I know many laggiful server owners will run in terror at the word "mobs", but I think that minetest can't say it's for old computers anymore. It's just too filled with features. If we wanna leave a version for older computers, someone better fork minetest before this gets added (if it does).

**Comment by Wuzzy2** [comment]
As for the fork: It already happend, it is called “Voxelands” now (formerly “Minetest-Classic”).

**Comment by falsechicken** [comment]
:thumbsup: I too would like to see faster mob processing at some point in the future.

**Comment by Wuzzy2** [comment]
Well, there are a lot of open questions: What should the “mob API” from Minetest do? Which capabilities do you wish (both of the API and the mobs themselves)? What features? How should/would mobs differ from entities? What Lua functions should be available? Which would be the limits? Do you want a more generic approach to mobs or a more specific one? What kinds of mobs should exist? How should mobs relate to the groups system? What kind of attacks would be possible? What could modders do if they find shortcomings in the mob API?

And, and, and …
Questions over questions.

**Comment by falsechicken** [comment]
Indeed. Tons of feedback required.

**Comment by nerzhul** [comment]
i also agree that the engine must have a little mob/pnj processor with high lua extensibility. But it needs to be discussed with other devs

**Comment by rubenwardy** [comment]
The problem is making it extendable.

Let's design it from the top down. It needs to support these types of mobs:
(please feel free to expand on this)

# Must do
* Wild Prey / Herbivores (sheep, pigs, cows, horses, deer, tuna)
    * Run away from predators and when hit
    * Can run away from players if timid
* Farm Animals (sheep, pigs, cows, horses)
    * Derived from wild prey, but like security and require players to feed them. Maybe. (Minecraft doesn't distinguish between wild prey and farm animals)
* Domesticated (cats and dogs from tigers and wolves)
   * one of my favourite things in MC are the dogs
* Predators (wolves, bears, sharks)
   * Different levels of aggressiveness. (will it attack players? will it attack prey? will it attack predators?)
* Monsters (zombies, 'creeper'-style, kraken 
    * Only attack players. All animals are scared of it.
* Animal Location
    * All of the above categories can be land, sea or air.

# Could do
* Breeding / selection - you can create more efficient animals by breeding them together.

# Why C++?
Client side prediction would be required to make it seamless on servers. C++ runs faster than Lua, less switching between C++ and Lua.

**Comment by nerzhul** [comment]
It's not the priority, because there are many things to do in the engine before adding new huge features for game, but we need to discuss on it together to find a good implementation

**Comment by rubenwardy** [comment]
The priorities should be:
* Faster protocols and more client side prediction
* Increasing user friendliness and experience, which includes removing blocking bugs.

**Comment by nerzhul** [comment]
agree

**Comment by falsechicken** [comment]
Man rubenwardy you go all in with everything you do don't you? I love it. No slackin' round here.

**Comment by rubenwardy** [comment]
It's only because I am on the computer and can type quickly. The thing is that it is all well and good suggesting a feature, but it would need to be implemented and methods of writing it needs to be discussed as well.

**Comment by nerzhul** [comment]
@rubenwardy agree

**Comment by C1ffisme** [comment]
Wow, so many comments in one day!

I like @rubenwardy 's ideas, and I think they would do nicely for an engine. See, what I'm thinking is we could have MobEntity already have some things made in C++ that LuaEntities must have in lua.

**Comment by SmallJoker** [comment]
I think Minetest shouldn't contain any mobs by default.
A mob API for mods would be a better idea. That way, it would be easy to create new mobs without editing Minetest's engine.

**Comment by rubenwardy** [comment]
That is what we are saying.

**Comment by rubenwardy** [comment]
Providing that mobs don't lag much on miltiplayer, I think they should be part of minetest_game, although they should be able to be disabled from a setting.

Explicitly disable subgame mods could be good for this.

**Comment by PilzAdam** [comment]
> I think they should be part of minetest_game, although they should be able to be disabled from a setting.

"Mod" is the word you are looking for here.

**Comment by rubenwardy** [comment]
No, because mods aren't counted as part of Minetest. They're mods.

**Comment by rubenwardy** [comment]
The whole "you can just install a mod" argument is damaging the gameplay experience. Minetest should be customisable, but it should already be shipped with a fully working and fun game. While I don't particularly care for monsters, animals add to the 'immersion' of the game.

**Comment by C1ffisme** [comment]
I agree with rubenwardy. We can't just make this game for developers. It turns out we've all been under the proggramming sun too long. There are those who hide in the gamer caves and...

Did I ever mention I write books for NaNoWriMo? (National Novel Writing Month).

Anyway, we can't assume everybody wants to get tons of mods. I know people who download and sometimes make mods for minecraft, but I'm sure there is a distinct amount of people who use vannila entirely. Meanwhile, you can't play minetest without mods. (Or using V7 and attempting a very explorative survival, but that uses lots of CPU.)

**Comment by Wuzzy2** [comment]
Please stay on topic, I thought we are just talking about the mobs API. DIscussion about minetest_game should go here:
* https://github.com/minetest/minetest_game
* https://forum.minetest.net/viewtopic.php?f=15&t=9724

Other Minetest-related stuff goes here:
* https://forum.minetest.net/viewforum.php?f=3

Not that I am a moderator, I'm just saying …


Anyways, back to topic:
I want to emphasize the following points on a mob API:
* There should be some kind of basic movement patterns
* The movement patterns should be treated independently from behaviour patterns (aggressive, passive, fleeing, etc.)
* It would be generally good if mobs would be technically not too terribly different from players, so features which mobs have (custom collision boxes, custom max. HP, etc.) should also be changable for players and vice-verser. This allows for more flexibility.

Examples for basic movement patterns:
* Static: The mob does not move at all and is not subject to gravity
* Static Faller: Same as static, but is also subject to gravity (probably useful for human-eating plants)
* Swimmer: The mob swims at the surface of some liquid
* Diver: The mob moves inside a liquid, may also be
* Flyer: The mob moves in the air or any other air-like non-liquid non-walkable node. But it cannot stand still.
* Walker: The mob walks on the ground like the player does
* Floater: The mob can move freely through air (or similar) and does not have to keep on moving
* Hybrid: Can combine two or more movement patterns (if feasible). It is unclear how, and which ones.
* Custom: Maybe, maybe, mods could add their own movement patterns. The question is: How?

I also would like to invite BlockMen into the discussion, because BlockMen authored the Creatures mod. I think we could get some good input from there.

@BlockMen

**Comment by nerzhul** [comment]
@C1ffisme agree, it's not only a developper playground it's a serious game used by players.
I have a good experience with NPCs on mangos emulator for WoW, i think we can tanke some parts of their approach

**Comment by rubenwardy** [comment]
Yes, it is offtopic. Minetest_game stuff should be worried about later.

The movement patterns would need to support any monster/animal you could
think of. I'm not sure about plants. They could be cool, but they would be
better as interactive nodes in my opinion, but that's a different feature.
Unless they are in plant pots and can jump. Birds can walk and can fly.

It would be useful to find some PDFs of papers that describe existing
algorithms, and look at how other projects do it. I can't search very well
on my phone.

**Comment by 4aiman** [comment]
Movement patterns may be added through Lua with C++ speed if there would be some algorithm we can pass as a table and there will be a "decipher" which will interpret the passed table as a patterns for every case: idle, running away, following smth/someone, getting punched, attacking, jumping, swimming, flying, etc

Those tables may even contain several concurrent patterns chosen pseudo-randomly (for client prediction to work seamless).


It would be also wise to add the ability to override C++ methods through Lua, but that is smth we already have. Can it be optimized by itself? (E.g. not because network proto became faster).

**Comment by hunter688** [comment]
Mobs are needed in the game, without them even in multiplayer is boring.

**Comment by C1ffisme** [comment]
Is anyone willing to work on a usable mob engine?

If we start a simple one, we could work up to something decent.

**Comment by sofar** [comment]
I'm not making any promises, but I have started work on a mob engine/framework/whatever.

I will share design goals and implementation later, when there is something to share. I'm using the time to teach myself the various skill that I need to make this successful, as well as do the design work without being encumbered by people telling me to make a pixel darker.

If you follow my youtube channel you can monitor progress. I'm regularly posting videos of what the current state is of the engine, although I'm explicitly not giving any details yet at this point in time.

I will not go into discussion about technical details in this thread. I have collected my requirements by looking at feedback from players on the forums and github, and am considering all the opinions already, so there's very little to be added. Needless to say that I have my standards set very high and I'm not making something simple (even though it may appear simple at first sight).

**Comment by C1ffisme** [comment]
@sofar Well, I hope something good comes out.

I think this issue should be closed, and a discussion on the forum opened.

**Comment by rubenwardy** [comment]
This is an often requested and important (imo) feature. This is a good place to discuss it

**Comment by nerzhul** [comment]
The first thing to do is to have a base class between all lifen SAO (PlayerSAO and LuaEntitySAO) which have common features, i propose UnitSAO as an intermediate which have some common features. I could provide a PR backported from my fork for this.

**Comment by neoascetic** [comment]
I guess the most problems with performance of different mobs mods is not C++ vs. Lua speed, but lacking of architecture. The same optimizations/algorithms may be done in Lua IMO as a mod as proof-of-concept and only later, when there are lot of good feedback from users and server owners, moved into the core and implemented in C++. This guarantees faster reaction on users request due dynamic nature of Lua an independence of engine's development process which leads to better understanding of what kind of API would be useful, and then periodically part of that API may be moved to the core.

**Comment by 4aiman** [comment]
> and only later, when there is a lot of good feedback from users and server owners, moved into the core and implemented in C++.

Then there should be an official public poll or else any new mobs API would end like "yet another mobs API". 
What I mean is, do we really have no good APIs for mobs at all?


I'd suggest making a new poll for every major release and add to C++ whichever API would be the most popular.

**Comment by sofar** [comment]
> I'd suggest making a new poll for every major release and add to C++ whichever API would be the most popular.

Minetest development (just like any OSS project) isn't a democracy.

I don't think the mob API should be in C++ *ever*. @neoascetic 's comment that likely this is best coded in Lua, and then optimized where needed in C++, is far more realistic and the right approach. Make it in Lua, optimize in C++ where needed.

If you look at my nodeupdate optimization, you can see that with well-written solutions, you really don't have to grab for C++ that quickly.

**Comment by C1ffisme** [comment]
I'm going to have to agree with @sofar.

Doing mobs in the engine would mean that a large chunk of the mob API would be hard-coded, and that wouldn't be very good for an engine like Minetest, where you can create anything you want.

**Comment by Ferk** [comment]
I agree that the problem is not a C++ vs Lua. However, I believe it's more of a client vs server issue. Mobs (and other entities like mounts, cars, boats and other transports) work flawlessly in singleplayer, where server-client latency is almost nothing.

The problem is that Lua is only server-side, so it's not possible to code movement patterns to be performed in the client (which would minimize lag). Also, having client-side logic might bring other problems like players getting out of sync and cheaters using modded clients if this logic is not monitored server-side as well periodically (which might end up resulting in latency issues anyway unless a certain margin of error was allowed).

**Comment by neoascetic** [comment]
#3440 is related. As I've said, having client-side Lua scripting would be a great improvement.

**Comment by MarkuBu** [comment]
> Is anyone willing to work on a usable mob engine?

I will start my own engine, but with a different approach than the others. I will write it in lua, but with C++ in mind. Each function, that should be written in C++ will be separated from the other parts 

I don't think, there is really a need for a huge mob api, because the common functions for all mobs can be reduced to a hand full of functions. Register, create, delete and move. And they already exist

But there are functions like a collision callback, which are totally missing yet and which should be implemented in C++. Also different kind of "sensors" with callbacks would be helpful to make mobs more complex and give them more abilities. But these sensors must be written in C++ because the necessary informations are not available for mods or only at a high cost and because of possible lacks of more than a second in functions like on_step or register_globalstep. 

The behaviour should be done server side, but in a script, because each mob can have different behaviours. Instead of putting different fixed behaviours into the api it would make more sense to provide some basic behaviours as templates which can be modified (or not)

I don't think, client side scripting solves the problem. It will open new issues with collision detection, out of sync problems between players and more. There should be other improvements in the client/server communication to solve this.

**Comment by C1ffisme** [comment]
I think there's a MTG issue about this, and I feel like this issue is kind of pointless now, since the issue discusses the implementation of mobs at all, which most definitely includes engine support.

Does anyone care if I close this?

**Comment by nerzhul** [comment]
don't close it as it's an issue we will solve, maybe in next release (0.5)

**Comment by C1ffisme** [comment]
Ok.

**Comment by paramat** [comment]
See MTGame issue https://github.com/minetest/minetest_game/issues/1667 especially sofar's comment https://github.com/minetest/minetest_game/issues/1667#issuecomment-290555777

**Comment by bell07** [comment]
Agree with @MarkuBu. At the time the whole mob logic is allways in on_step called unreasonable multiple times each second. Each call the method gets environment data and entity data, check if anything is happened and react on them. That is the main impact at the time. The entity should pass events from c++ world to the lua callbacks. 
1. The most important is "on_collision". Should be called if entity collides an other entity or an solid node. Using this callback an arrow mod can be implemented without any "steps". Just set velocity, acceleration and on_collision callback in the mod for example. 
2. Other callbacks: on_before_chasm_reached?

3. entity:navigate_to(pos or entity). An path calculator already exists in engine. A method is required to let an entity to follow the path, with "on_direction_change" and "on_stuck" callback. Often implemented method is to follow an player or an other entity. So the path calculator should be able to adjust the path if "destination entity" moves.

4. The entity's are blind at the time. The mos mobs-mods implements "core.get_node" or VoxelManip methods to get world data. Disadvantage is the data is copied. The engine should be able to provide an VoxelManip-like area around the entity without copying data from map to lua table or any other perfomant way to do it (optional).

**Comment by Wuzzy2** [comment]
> The entity's are blind at the time. The most mobs-mods uses "core.get_node" or VoxelManip methods to get world data. Disadvantage at this point: the data is copied. The engine should be able to provide an VoxelManip-like area around the entity without copying data from map to lua table or in any other perfomant way to do it.

:+1: so much. This will also be incredibly useful for non-mob entities. Using `minetest.get_node` or even a LuaVoxelManip feels incredibly clunky and wrong on so many levels, especially if you need this info basically at every step.
The point here is not that this isn't already possible, but that it just seems the wrong way.
If there would be a standard and efficient (!) way to get the immediate surroundings of an entity, this would be a great improvement for all entities, really. Not just mobs. It will help for boats, minecarts, flying carpets ( :D ), and much more.

Fun fact: Currently, falling node entities query `minetest.get_node` each step.

**Comment by bell07** [comment]
> Fun fact: Currently, falling node entities query minetest.get_node each step.
can be improved in "blind way" by on_collision callback, if implemented :D

**Comment by Mizux** [comment]
Why closing this while other duplicate have been also closed ? i.e. where the entity/mob feature is tracked ?
* https://github.com/minetest/minetest/issues/7608
* https://github.com/minetest/minetest_game/issues/1667
* https://github.com/minetest/minetest_game/issues/2710

If minetest don't want to provide any entity support, would be happy to have it written...

**Comment by valenciaDev74** [comment]
> Games and mods can already implement mobs, I don't see any specific need for an engine API here. This issue has also been open for a long time with no real conclusion
> 
> What would be good is more engine support for features that makes mobs mods easier/better, that's something that should be created as new issues. Examples include pathfinding improvement, animations, and sensing utilities

I think the complaint is more for the results than for the development experience, I have an old pc with 4gb of ram, 2.90GHz and nvidia graphics integrated with 1657mb and the truth is that minecraft mobs tremendously outperform those of minetest, I mean it feels good not to have cows flying when they touch the water or birds completely glitch every time they move.

**Comment by bell07** [comment]
I think this issue should be reopened and renamed into "Enhance entity API to support mobs mods development".

In my last steps in mobs development attempts I missed the next entity methods:

- on_collision()
Currently the on_step is used to poll the position each step that is not always performant or accurately.

- navigate_to(pos)
This method should cover the path calculations and be able to navigate the mob to position according his "physical" attributes (to be defined). Navigation should run without on_step() utilization but can be stopped from mod in on_step() or on_collision() .

But now I scrolled up and see I wrote the same in [2017]( https://github.com/minetest/minetest/issues/2291#issuecomment-348602939)

**Comment by appgurueu** [comment]
`on_collision` seems like a sane feature. I agree that `moveresult` in `on_step` is suboptimal. That should get a separate issue rather than "repurposing" this ancient issue. Feel free to open a new one.

`navigate_to` is much, much trickier. The pathfinding part should already be covered by https://github.com/minetest/minetest/issues/5479. It also requires many specific choices be made that I think shouldn't be hardcoded, so I'm wary of trying to design such an API that fits everyone's use cases.

**Comment by rubenwardy** [comment]
Please open new tickets for those features
